[Compose 페이먼츠 3단계] 다이스 미션 제출합니다.

moondev03 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
```
3단계 미션 리뷰도 잘 부탁드립니다 🙇‍♂️
이전 단계에서 주신 리뷰도 최대한 반영했습니다 :)
```

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

[payments_step_3.webm](https://github.com/user-attachments/assets/46b42123-5578-4195-b8f4-1060dc32d2ec)

moondev03 committed:
build: Compose lint checks 추가

- `com.slack.lint.compose:compose-lint-checks` 라이브러리 의존성을 추가합니다.

moondev03 committed:
refactor: `coerceAtMost`를 `coerceInLength`로 변경

moondev03 committed:
refactor: 도메인 모델의 Error를 Exception으로 이름 변경

- `CardPasswordError` -> `CardPasswordException`
- `CardNumberError` -> `CardNumberException`
- `CardExpirationDateError` -> `CardExpirationDateException`
- `CardholderNameError` -> `CardholderNameException`

moondev03 committed:
refactor: 도메인 모델의 예외 타입 변경에 따른 ViewModel 수정

moondev03 committed:
feat: `PaymentCardsUiModel` 추가

moondev03 committed:
refactor: 오류 메시지 표시 로직 수정

- `CardRegistrationScreenUiState`에서 각 입력 필드의 오류 메시지를 결정하는 로직을 수정했습니다.

moondev03 committed:
refactor: ViewModel에 Saver 구현 및 UiEvent에서 None 상태 제거

moondev03 committed:
refactor: PaymentCard가 UiModel 대신 원시 값을 받도록 변경

moondev03 committed:
style: PaymentCardText 컴포저블의 스타일 속성을 매개변수로 추출

- `PaymentCardText` 함수에서 하드코딩된 스타일 속성(color, fontSize, letterSpacing, lineHeight, fontWeight)을 매개변수로 변경하여 재사용성을 높였습니다.

moondev03 committed:
refactor: ViewModel에서 uiEvent 속성의 getter 제거

moondev03 committed:
refactor: 카드 소유자 이름 유효성 검사 로직 개선

moondev03 committed:
style: PaymentCardText의 letterSpacing 기본값 변경 및 maxLines 추가

- `PaymentCardText`의 `letterSpacing` 기본값을 `1.sp`에서 `0.sp`로 수정
- 텍스트가 한 줄로 표시되도록 `maxLines = 1` 속성 추가

moondev03 committed:
docs: 3단계 미션 README 추가

moondev03 committed:
feat: 카드사 선택을 위한 리소스 추가

- `BankTypeUiModel` enum 클래스 추가
  - 카드사별 배경색, 이름, 로고 리소스 ID 정의
- `strings.xml`에 카드사 이름 추가
- `drawable`에 카드사 로고 이미지 추가

moondev03 committed:
feat: 카드사 선택 바텀시트 UI 구현

- 카드사 로고와 이름을 표시하는 `BankSelectRow` 컴포저블 추가
- `BankSelectRow`를 포함하고 `ModalBottomSheet`를 사용하여 은행 선택 UI를 제공하는 `BankSelectBottomSheet` 컴포저블 추가

moondev03 committed:
feat: 카드 UI에 은행 이름 표시 및 디자인 수정

- `PaymentCard` 컴포저블에 은행 이름을 표시하도록 `bankName` 매개변수를 추가했습니다.
- 카드 컴포넌트의 내부 여백(padding)을 조정했습니다.
- 은행 이름과 카드 번호의 글자 크기 및 자간을 수정하여 가독성을 개선했습니다.

moondev03 committed:
feat: 카드사 선택 기능 추가

- 카드 등록 화면 진입 시 카드사 선택 BottomSheet 표시
- 카드사 선택 시 화면에 선택된 카드사 정보 반영
- `PaymentCardUiModel`에 카드사 정보(`bankType`) 추가
- `PaymentCard` 컴포저블에 카드사 이름 및 배경색 표시 로직 추가

moondev03 committed:
refactor: PaymentCard 레이아웃 Column에서 Box로 변경

- `Column`과 `Arrangement.SpaceBetween`을 사용하는 대신 `Box`와 `align` 수식어를 사용하여 레이아웃을 구성합니다.

moondev03 committed:
feat: 카드 이미지를 클릭하여 은행 선택 초기화 기능 추가

- 카드 등록 화면의 카드 이미지에 클릭 이벤트 추가
- 클릭 시 선택된 은행 정보(`bankType`)를 초기화(`NOT_SELECTED`)하도록 ViewModel의 `updateBank` 함수를 호출

moondev03 committed:
refactor: 은행 선택 BottomSheet 상태 관리 로직 추가

- `CardRegistrationScreenViewModel`에 BottomSheet를 열고 닫는 함수 추가
- `BankSelectBottomSheet`에 `onDismiss` 콜백 추가 및 호출 로직 구현
- `CardRegistrationScreen`에서 카드 클릭 시 BottomSheet를 열도록 수정
- `CardRegistrationScreenUiState`에서 `openBankSelectorBottomSheet` 상태를 직접 관리하도록 변경

moondev03 committed:
refactor: LiveData observeAsState의 by delegate를 value로 변경

- `CardsScreen`과 `CardRegistrationScreen`에서 `observeAsState`를 사용할 때, by delegate 대신 `.value`를 직접 사용하도록 수정했습니다.
- `uiState`가 null일 경우 Composable이 렌더링되지 않도록 early return을 추가했습니다.
- `uiEvent` 처리 시 불필요한 지역 변수 할당을 제거했습니다.

moondev03 committed:
refactor: CardRegistrationButtonPreview를 private으로 변경

moondev03 committed:
docs: 3단계 기능 요구 사항 반영

- 카드 추가 화면에서 카드사 선택 필수 기능 구현
- 선택된 카드사에 따라 카드 미리보기 변경 기능 구현
- 카드사 선택 시 아이콘 노출 기능 구현
- `FlowRow`와 `ModalBottomSheet`를 활용하여 카드사 선택 기능 구현

moondev03 committed:
test: 카드 정보 입력 및 화면 표시 테스트 수정

- `PaymentCard` 테스트 로직 수정
  - UI 모델 대신 개별 속성(은행명, 카드번호 등)을 전달하여 테스트하도록 변경
  - 카드사 이름, 카드번호, 만료일, 소유자 이름이 올바르게 표시되는지 확인
- 입력창(TextField) 테스트 이름 구체화
  - `에러메시지가_있으면` -> `오류_메시지가_제공되면`
  - `값을_입력하면` -> `(카드번호/만료일 등)_입력이_있으면`
- 유효한 카드 정보 fixture에 `bankType` 추가

moondev03 committed:
test: 카드사 선택 화면 테스트 추가

- `BankSelectRow` 테스트 추가
  - 카드사 목록이 표시되는지 확인
  - 카드사 선택 시 `onBankSelected` 콜백이 호출되는지 확인
- `BankSelectorBottomSheet` 테스트 추가
  - 카드사 선택 시 `onBankSelected`와 `onDismiss` 콜백이 모두 호출되는지 확인

moondev03 committed:
feat: 접근성을 위한 contentDescription 추가

- PaymentCard 내 카드 정보(카드사, 번호, 만료일, 소유자 이름)에 contentDescription을 추가했습니다.
- 카드사 선택 바텀 시트(BankSelectorBottomSheet)에 contentDescription을 추가했습니다.

moondev03 committed:
test: 카드사 선택 바텀 시트 테스트 추가

- 화면 진입 시 카드사 선택 바텀 시트가 표시되는지 확인
- 바텀 시트에서 카드사를 선택했을 때, 해당 카드사 이름이 카드 미리보기에 표시되는지 확인

moondev03 committed:
refactor: 람다를 멤버 참조로 변경

moondev03 committed:
refactor: `isDigitsOnly` 확장 함수 추가 및 적용

- `String.isDigitsOnly` 확장 함수를 추가하여 도메인 모델의 안드로이드 프레임워크 의존성을 제거했습니다.

moondev03 committed:
refactor: IntRangeExt 파일을 삭제하고 PaymentCardUiModel로 함수 이동

- `IntRange.kt` 확장 파일을 삭제하고, 해당 파일에 있던 `coerceInLength` 함수를 `PaymentCardUiModel` 클래스 내부의 private 함수로 이동

moondev03 committed:
refactor: `isSaveButtonEnabled`를 `canRegisterCard`로 이름 변경

moondev03 committed:
fix: isDigitsOnly()가 숫자로만 구성되었는지 정확히 확인하도록 수정

moondev03 committed:
test: `assert`를 `assertEquals`로 변경

moondev03 committed:
refactor: BankSelectRow의 contentDescription에 toBankName

moondev03 committed:
refactor: PaymentCard 컴포저블의 contentDescription 통합

moondev03 committed:
refactor: 카드 소유자 이름의 최대 길이를 상수로 설정

- `CardholderNameUiModel`의 `maxLength` 기본값을 `CardholderName.MAX_LENGTH` 상수로 변경

moondev03 committed:
feat: 카드 등록 가능 조건에 은행 선택 여부 추가

moondev03 committed:
refactor: CardNumberException, CardPasswordException을 data class로 변경

moondev03 committed:
refactor: BankTypeUiModel에 어노테이션 추가

- `nameResId`에 `@StringRes` 어노테이션을 추가했습니다.
- `logoResId`에 `@DrawableRes` 어노테이션을 추가했습니다.

moondev03 committed:
fix: CardRegistrationScreen 미리보기에 ViewModel 추가

- `CardRegistrationScreenPreview`에 `CardRegistrationScreenViewModel`의 인스턴스를 추가하여 미리보기가 정상적으로 동작하도록 수정

giovannijunseokim commented on app/src/androidTest/java/woowacourse/payments/ui/component/BankSelectorBottomSheetTest.kt:
```diff
@@ -0,0 +1,43 @@
+import androidx.compose.ui.test.assertIsDisplayed
+import androidx.compose.ui.test.junit4.createComposeRule
+import androidx.compose.ui.test.onNodeWithText
+import androidx.compose.ui.test.performClick
+import org.junit.Rule
+import org.junit.Test
+import org.junit.jupiter.api.assertAll
+import woowacourse.payments.ui.model.BankTypeUiModel
+
+class BankSelectorBottomSheetTest {
+    @get:Rule
+    val composeTestRule = createComposeRule()
+
+    @Test
+    fun `카드사를_선택하면_선택된_카드사와_닫기를_전달한다`() {
+        // given
+        var selectedBank: BankTypeUiModel? = null
+        var dismissed = false
+
+        composeTestRule.setContent {
+            BankSelectBottomSheet(
+                onBankSelected = { selectedBank = it },
+                onDismiss = { dismissed = true },
+            )
+        }
+
+        // when
+        composeTestRule
+            .onNodeWithText("국민카드")
+            .assertIsDisplayed()
+            .performClick()
+
+        // then
+        composeTestRule.waitForIdle()
+
+        assertAll(
+            { assert(selectedBank == BankTypeUiModel.KB) },
+            { assert(dismissed) },
+        )
+    }
```
하나의 테스트에서 여러 assert문이 쓰이고 있습니다.
이렇게 작성했을 때의 문제점은 없을까요? 👀

moondev03 replied:
- 카드사 선택 시 선택된 카드사 전달 확인 테스트
- 카드사 선택 시 닫기 이벤트 전달 확인 테스트

2개의 케이스로 분리하여 테스트하도록 변경했습니다.

giovannijunseokim commented on app/src/main/java/woowacourse/payments/domain/CardExpirationDate.kt:
```diff
@@ -10,30 +10,43 @@ value class CardExpirationDate private constructor(
 
-        class NonDigit : CardExpirationDateError("만료일은 숫자만 입력할 수 있습니다.")
+        data object NonDigitException : CardExpirationDateException("만료일은 숫자만 입력할 수 있습니다.") {
+            private fun readResolve(): Any = NonDigitException
+        }
 
-        class UnsupportedDate : CardExpirationDateError("존재하지 않거나 유효하지 않은 만료일입니다.")
+        data object UnsupportedDateException :
+            CardExpirationDateException("존재하지 않거나 유효하지 않은 만료일입니다.") {
+            private fun readResolve(): Any = UnsupportedDateException
+        }
```
`readResolve`는 사용되지 않고 있는 것 같아요. 무엇을 위한 함수인가요? 👀

moondev03 replied:
`readResolve()`는 직렬화 과정에서 싱글톤 객체의 유일성을 보장하기 위한 함수입니다.  

- `object`나 `data object`는 프로그램 내에서 단 하나의 인스턴스를 가지지만,  직렬화 후 역직렬화 시 JVM이 새로운 객체를 생성할 수 있어 싱글톤 특성이 깨질 수 있습니다.
- `readResolve()`를 구현하면 역직렬화 시 항상 기존 싱글톤 인스턴스를 반환하도록 보장할 수 있습니다.

다만, 현재 미션에서는 해당 객체를 직렬화할 일이 없기에 실제로 사용되지는 않지만 IDE에서 주의 메시지가 노출되기에 작성했습니다.

giovannijunseokim commented on app/src/main/java/woowacourse/payments/domain/CardExpirationDate.kt:
```diff
@@ -10,30 +10,43 @@ value class CardExpirationDate private constructor(
-        class ExceedsLength : CardExpirationDateError("만료일은 ${VALID_CARD_EXPIRATION_DATE_LENGTH}자리여야 합니다. (입력값이 너무 깁니다)")
-
-        class InsufficientLength : CardExpirationDateError("만료일은 ${VALID_CARD_EXPIRATION_DATE_LENGTH}자리여야 합니다. (입력값이 너무 짧습니다)")
+        class InvalidLengthException(
+            val kind: Kind,
+        ) : CardExpirationDateException("만료일은 ${VALID_CARD_EXPIRATION_DATE_LENGTH}자리여야 합니다.") {
+            enum class Kind { INSUFFICIENT, EXCEEDS }
+        }
```
`Exception`과 상속 계층을 통해 세세히 분기처리 해주셨습니다.
다만 자연스레 코드 복잡도가 많이 올라가고 가독성도 떨어진 것 같아요. 🤔 

해당 방식을 구현해보고 느낀 장점, 단점이 무엇인가요? 👀

moondev03 replied:
검증 실패 시 사용자에게 적절한 피드백을 제공하고, 시스템적으로 안전하게 처리하기 위해서는 해당 값이 왜 유효하지 않은지 명확히 구분할 수 있어야 한다고 생각합니다.

저 역시 지오가 느끼는 것 처럼 코드의 복잡도가 올라가고 그로 인해 가독성이 떨어지는 것을 인지하고 있습니다.

그렇다면, 이런 비즈니스 로직을 검증할 때 지오는 어떻게 처리하는 것이 적절하다고 보시나요?

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/BankSelectorBottomSheet.kt:
```diff
@@ -0,0 +1,58 @@
+@Composable
+fun BankSelectBottomSheet(
+    onBankSelected: (BankTypeUiModel) -> Unit,
+    onDismiss: () -> Unit,
+    modifier: Modifier = Modifier,
+) {
+    val sheetState = rememberModalBottomSheetState(confirmValueChange = { false })
+    var selectedBank by remember { mutableStateOf(BankTypeUiModel.NOT_SELECTED) }
+
+    LaunchedEffect(selectedBank) {
+        if (selectedBank != BankTypeUiModel.NOT_SELECTED) {
+            sheetState.hide()
+            onDismiss()
+        }
+    }
```
`onDismiss`라는 인자명을 갖고 있지만,
실제 수행되는 상황은 `Dismiss될 때` 보다는 `은행이 선택됐을 때`에 더 걸맞는 것 같아요. 👀

moondev03 replied:
동의합니다.
상황에 맞는 콜백 메서드를 호출하도록 변경했어요 :)

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/component/PaymentCard.kt:
```diff
@@ -40,32 +46,42 @@ fun PaymentCard(
         modifier = modifier,
+        maxLines = 1,
     )
 }
```
`text`만 필수 인자라는 점에서 `Text` 컴포저블과 매우 유사해보이는데요. 👀 

이렇게 하면 기본 `Text` 컴포저블을 사용하는 것에 비해 어떤 이점이 있을까요? 🤔

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/model/BankTypeUiModel.kt:
```diff
@@ -0,0 +1,63 @@
+        null,
+    ),
+    BC(
+        Color(0xFFFA3246),
```
위의 `Color.DarkGray`처럼 추출해본다면 어떤 장점이 있을까요? 👀

moondev03 replied:
혹시 모든 색상을 추출하는 게 필요하다고 보시나요?

현재 카드 색상은 고유하고 재사용되지 않으며, 테마 대응이 필요하지 않기 때문에 굳이 추출할 이유가 없어 보다고 생각했어요.

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreen.kt:
```diff
@@ -10,69 +11,74 @@ import androidx.compose.foundation.verticalScroll
-                    ?.let(onRegistrationComplete)
-            }
+            is CardRegistrationScreenUiEvent.RegisteredCard ->
+                uiEvent.paymentCard.let(onRegistrationComplete)
```
스코프 함수 `let`이 어떤 의도로 사용된 것일까요? 👀

moondev03 replied:
앗, 실수

[ac100cb](https://github.com/woowacourse/android-payments/pull/62/commits/ac100cba2940f31c257a0b95888c1a33aa840144)

giovannijunseokim commented on app/src/main/java/woowacourse/payments/ui/screen/registration/CardRegistrationScreenUiState.kt:
```diff
@@ -14,44 +15,57 @@ import woowacourse.payments.ui.model.PaymentCardUiModel
+            -> null
         }
 
     fun toPaymentCardUiModel(): PaymentCardUiModel =
```
다이스가 생각하는 `UiState`와 `UiModel`의 차이는 무엇인가요? 👀

moondev03 replied:
---
### UiModel
- UI가 요구하는 데이터를 가공/변환
- UI 로직과 무관하게 표현용 데이터 구조를 제공

### UiState
- UI가 그려지는 전체 상태를 캡슐화
- 화면에서 보여줄 상태를 묶어 관리하는 객체

---
즉, UiModel은 UI가 필요로 하는 데이터 자체, UiState는 화면 전체의 상태와 흐름을 관리한다는 차이가 있다고 생각해요

giovannijunseokim reviewed (changes requested):
안녕하세요 다이스. 미션 구현 고생하셨습니다! 👍 

궁금한 점들을 리뷰 내용으로 남겨두었으니 확인 후 답변 부탁드립니다. 😁 

궁금한 점이나 논의하고 싶은 내용이 있으면 편하게 얘기해주세요! 🙏

moondev03 committed:
test: `BankSelectorBottomSheet` 테스트 분리

- 카드사 선택 시 선택된 카드사 전달 확인 테스트 추가
- 카드사 선택 시 닫기 이벤트 전달 확인 테스트 추가

moondev03 committed:
refactor: 카드사 선택 BottomSheet 개선

- 시스템 뒤로가기로 인한 hide 처리 시 onDismissRequest 콜백 처리
- 네이밍 개선

moondev03 committed:
refactor: PaymentCard Text 스타일을 LocalTextStyle로 제공

- PaymentCardText 컴포저블 함수 제거
- paymentCardTextStyle 함수 추가 및 LocalTextStyle을 통해 하위 컴포저블에 스타일 제공
- Text 컴포저블의 maxLines 속성 추가

moondev03 committed:
refactor: `PaymentCardInfoBlock` 컴포저블 함수명 변경

- `PaymentCardInfoBlock`을 `PaymentCardDetails`로 변경

moondev03 committed:
refactor: `PaymentCardInfoBlock` 컴포저블 함수명 변경

- `PaymentCardInfoBlock`을 `PaymentCardDetails`로 변경

moondev03 committed:
refactor: `let` 함수를 `run` 함수로 변경

giovannijunseokim reviewed (approved):
리뷰 반영 고생하셨습니다! 🔥 

이번 미션은 여기서 머지하겠습니다. 🙇 

다음 미션도 응원하겠습니다! 👍
